### PR TITLE
Port `synth_clifford_greedy` to Rust

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

Addresses #12235.

This PR ports the method `synth_clifford_greedy` for synthesizing `Clifford` operators to Rust. In the process the algorithm was additionally improved by (1) ignoring all (expensive!) clifford phase computations till the very end, thus working on the "symplectic matrix" part of the tableau, (2) completely avoiding (expensive!) clifford `evolve` and clifford `adjoint` operations by pre-composing rather than post-composing the current Clifford with elementary Cliffords (corresponding to CX, SWAP, S, and H gates), (3) almost completely avoiding copying clifford objects, (4) fixing an inefficiency in cost computations.

**Runtime improvements:**
On my laptop, the old python code takes about *400* seconds to synthesize 50 random 50-qubit Cliffords, and the new code takes *0.25* seconds to do the same. So it's at least 1000x improvement in performance.

**Quality improvements:**
After fixing an inaccuracy in cost computations, the number of SWAPs introduced by the new algorithm is smaller than before; the totals for the 50 random cliffords from before are:
```
total_swap_old = 1688, total_cx_old = 63179
total_swap_new = 418, total_cx_new = 63283
```

### CODE REVIEW

**Comment by ShellyGarion** [comment]
it will be better to merge #12456 and #12598 and #12560 before this PR

**Comment by alexanderivrii** [comment]
This PR should be rebased on top #12456 for the computation of inverse binary matrices, Other than that, this is ready for review.

**Comment by alexanderivrii** [comment]
This is ready for review (@ShellyGarion, @Cryoris). All suggestions to make the code even faster or more "rusty" are highly welcome.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+
+use indexmap::IndexSet;
+use ndarray::{azip, s, Array1, Array2, ArrayView2};
+use smallvec::{smallvec, SmallVec};
+
+use crate::synthesis::linear::utils::calc_inverse_matrix_inner;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::Qubit;
+
+/// Symplectic matrix.
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs:21]: I would suggest to move all this code (until line 232) to a helper file called `clifford_utils.rs`, which can be moved later to `quantum_info` when porting the `Clifford` class to rust (#12367).
We can change the API later, but we should also keep in mind that these classes would later be used for the general Clifford class operations.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+/// Currently this class is internal to the greedy synthesis algorithm.
+struct SymplecticMatrix {
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs:23]: I would suggest to move all this code (until line 232) to a helper file called clifford_utils.rs, which can be moved later to quantum_info when porting the Clifford class to rust (https://github.com/Qiskit/qiskit/issues/12367).
We can change the API later, but we should also keep in mind that these classes would later be used for the general Clifford class operations.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 1 above)
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: I would suggest to move all this code (until line 232) to a helper file called clifford_utils.rs, which can be moved later to quantum_info when porting the Clifford class to rust (https://github.com/Qiskit/qiskit/issues/12367).
We can change the API later, but we should also keep in mind that these classes would later be used for the general Clifford class operations.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+}
+
+/// Clifford.
+/// Currently this class is internal to the greedy synthesis algorithm
+/// and has a very different functionality from Qiskit's python-based
+/// Clifford class.
+struct Clifford {
+    /// Number of qubits.
+    num_qubits: usize,
+    /// Matrix with dimensions (2 * num_qubits) x (2 * num_qubits + 1).
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: what's the reason to have the phases as the last line of the matrix and not as a separate vector (unlike the current Clifford class)? will it be useful later?

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+    }
+}
+
+fn pauli_pair_to_index(xs: bool, xd: bool, zs: bool, zd: bool) -> usize {
+    ((xs as usize) << 3) | ((xd as usize) << 2) | ((zs as usize) << 1) | (zd as usize)
+}
+
+/// The five classes of Pauli operators as described in the paper.
+#[derive(Clone, Copy)]
+enum PauliClass {
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: the name is a bit confusing with the `Pauli` class, maybe change to `PauliPairsClass` ?

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+    ClassA,
+    ClassB,
+    ClassC,
+    ClassD,
+    ClassE,
+}
+
+/// The 16 pairs of Pauli operators are divided into 5 equivalence classes
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: Pauli 2-qubit operators

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+        adjust_final_pauli_gates(&mut clifford_gates, self.tableau, self.num_qubits)?;
+
+        Ok((self.num_qubits, clifford_gates))
+    }
+}
+
+/// Given a sequence of Clifford gates that correctly implements the symplectic matrix
+/// of the target clifford tableau, adds the Pauli gates to also match the phase of
+/// the tableau.
+fn adjust_final_pauli_gates(
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: this function could also be useful to synthesize a Clifford for LNN later

File: releasenotes/notes/oxidize-synth-clifford-greedy-0739e9688bc4eedd.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+upgrade_synthesis:
+  - |
+    The function :func:`.synth_clifford_greedy` that synthesizes :class:`.Clifford` operators
+    was ported to Rust.
```
Comment (ShellyGarion) [inline releasenotes/notes/oxidize-synth-clifford-greedy-0739e9688bc4eedd.yaml]: it's worth to elaborate here on the performance improvement

**Review by ShellyGarion — COMMENTED** [review]
The code looks great! My main comment is that we should think a little bit on the API, since it will be used for other Clifford synthesis methods as well as to port the Clifford class to rust later (#12367)

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 1 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: I am not completely sure what the real Clifford class will look like, so I have deliberately kept this internal for now. Let us change this when we start moving more Clifford-related stuff into Rust. I have also added a comment in [0d89a35](https://github.com/Qiskit/qiskit/pull/12601/commits/0d89a35b49b42f610c2f380f70d1185016c8c73c).

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 3 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: Hmm, this is an interesting suggestion, which has its pros and cons. The current form is consistent with how we store a Clifford tableau in Python, but we can change it of course. It is also easier to perform row operations directly on the (2n) x (2n+1) matrix.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 6 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: Sure! We will need to reorganize the code structure a bit once more Clifford synthesis algorithms are implemented.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 1 above)
Comment (ShellyGarion) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: ok, but I would still prefer that you move the functions and classes that are not specific to the greedy compiler to another file, say `clifford_utils.rs` or `utils.rs`

Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: Done in [4e9b00d](https://github.com/Qiskit/qiskit/pull/12601/commits/4e9b00d7b27584e0d5ba28ec18bf7832e15fbabe): the "common" code (`SymplecticMatrix`, `Clifford`, `adjust_final_pauli_gates`) is moved to `clifford/utils.rs`. For now these classes are still local to the synthesis library.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 4 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: renamed in [bbc6ceb](https://github.com/Qiskit/qiskit/pull/12601/commits/bbc6ceb158cd537dfe5f6909932132a75e51029b)

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 5 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: improved comment in [bbc6ceb](https://github.com/Qiskit/qiskit/pull/12601/commits/bbc6ceb158cd537dfe5f6909932132a75e51029b)

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 6 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs]: done in [bbc6ceb](https://github.com/Qiskit/qiskit/pull/12601/commits/bbc6ceb158cd537dfe5f6909932132a75e51029b)

File: releasenotes/notes/oxidize-synth-clifford-greedy-0739e9688bc4eedd.yaml (hunk 7 above)
Comment (alexanderivrii) [reply releasenotes/notes/oxidize-synth-clifford-greedy-0739e9688bc4eedd.yaml]: Done in [441c9f3](https://github.com/Qiskit/qiskit/pull/12601/commits/441c9f3a1a312320fdb2697a36073a1e16ad532d).

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+use indexmap::IndexSet;
+use ndarray::{s, ArrayView2};
+use smallvec::smallvec;
+
+use crate::synthesis::clifford::utils::CliffordGatesVec;
+use crate::synthesis::clifford::utils::{adjust_final_pauli_gates, SymplecticMatrix};
+use qiskit_circuit::operations::StandardGate;
+use qiskit_circuit::Qubit;
+
+fn pauli_pair_to_index(xs: bool, xd: bool, zs: bool, zd: bool) -> usize {
```
Comment (ShellyGarion) [inline crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs:27]: it's worth to explain this function: what are xs, xd, zs, zd?

**Review by ShellyGarion — COMMENTED** [review]
LGTM except a minor comment.
The implementation and correctness of the algorithm look good to me.
I would be happy if someone could also look into the rust implementation aspects.

File: crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs (hunk 8 above)
Comment (alexanderivrii) [reply crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs:27]: done!

**Review by ShellyGarion — APPROVED** [review]

**Pull request merged**